tools/heir-opt.cpp

```diff
@@ -155,6 +157,72 @@ void polynomialToLLVMPipelineBuilder(OpPassManager &manager) {
   manager.addPass(createSymbolDCEPass());
 }
 
+void tosaToBooleanTfhePipeline(const std::string &yosysFilesPath,
+                               const std::string &abcPath) {
+  PassPipelineRegistration<YosysOptimizerPipelineOptions>(
+      "tosa-to-boolean-tfhe", "Arithmetic modules to boolean tfhe-rs pipeline.",
+      [yosysFilesPath, abcPath](OpPassManager &pm,
+                                const YosysOptimizerPipelineOptions &options) {
+        // Secretize inputs
+        pm.addPass(createSecretize());
+
+        // TOSA to arith pipeline
+        // TOSA to linalg
+        pm.addNestedPass<FuncOp>(createTosaToLinalgNamed());
+        pm.addNestedPass<FuncOp>(createTosaToLinalg());
+        pm.addNestedPass<FuncOp>(createTosaToArith(true, false));
+        pm.addNestedPass<FuncOp>(createTosaToTensor());
+        pm.addPass(bufferization::createEmptyTensorToAllocTensorPass());
+        pm.addNestedPass<FuncOp>(createLinalgDetensorizePass());
+        pm.addPass(createConvertTensorToLinalgPass());
+        pm.addPass(bufferization::createEmptyTensorToAllocTensorPass());
+        // Bufferize
+        pm.addNestedPass<FuncOp>(createLinalgBufferizePass());
+        pm.addNestedPass<FuncOp>(tensor::createTensorBufferizePass());
+        pm.addPass(arith::createArithBufferizePass());
+        pm.addPass(func::createFuncBufferizePass());
+        pm.addNestedPass<FuncOp>(
+            bufferization::createFinalizingBufferizePass());
+        // Affine
+        pm.addNestedPass<FuncOp>(createConvertLinalgToAffineLoopsPass());
+        pm.addNestedPass<FuncOp>(memref::createExpandStridedMetadataPass());
+        pm.addNestedPass<FuncOp>(affine::createAffineExpandIndexOpsPass());
+        pm.addNestedPass<FuncOp>(memref::createExpandOpsPass());
+        pm.addNestedPass<FuncOp>(affine::createSimplifyAffineStructuresPass());
+        pm.addPass(memref::createFoldMemRefAliasOpsPass());
+        pm.addPass(createExpandCopyPass());
+        // Cleanup
+        pm.addPass(createMemrefGlobalReplacePass());
+        arith::ArithIntNarrowingOptions arithOps;
+        arithOps.bitwidthsSupported = {4, 8, 16};
+        pm.addPass(arith::createArithIntNarrowing(arithOps));
+        pm.addPass(createCanonicalizerPass());
+        pm.addPass(createSCCPPass());
+        pm.addPass(createCSEPass());
+        pm.addPass(createSymbolDCEPass());
+
+        // Wrap with secret.generic and then distribute-generic.
+        pm.addPass(createWrapGeneric());
+        auto distributeOpts = secret::SecretDistributeGenericOptions{};
+        distributeOpts.opsToDistribute = {"affine.for", "affine.load",
+                                          "affine.store", "memref.get_global"};
+        pm.addPass(secret::createSecretDistributeGeneric(distributeOpts));
+        pm.addPass(createCanonicalizerPass());
+        pm.addPass(secret::createSecretCaptureGenericAmbientScope());
```

Comment:
So I need this also to happen inside of YosysOptimizer, but I need it mid-pass. What do you think about just adding it to YosysOptimizer by default?